### ISSUE TITLE
Milestones week numbers don't match Timeline week numbers

### ISSUE THREAD

**Issue by ragesoss** [WikiEducationFoundation/WikiEduDashboard#2461]
In the Milestones component, the numbering of the weeks does not account for either `weeksBeforeTimeline` or empty weeks, so the numbering will not match the corresponding weeks on the Timeline tab if the timeline doesn't start on Week 1 of the course or if there are empty weeks before the last milestone.

The easiest solution is probably to duplicate the approach used in Timeline to iterate through the weeks and weekMeetings and add to the week number for each empty week.

The better solution that avoids duplicating the logic and spreading out this confusing bit of code to multiple locations is probably to move the whole process of week numbering to a selector, so that the logic only lives in one place *and* we don't have to duplicate the calculation each time the user changes tabs.

**Comment by khyatisoneji** [comment]
Can you please point to the code in timeline component that takes care of handling weeks and weekMeetings.
Thanks!

**Comment by ragesoss** [comment]
weekMeetings gets passed down from TimelineHandler, and most of the logic is in the render method of Timeline, lines 178 to 258.

**Comment by khyatisoneji** [comment]
Will take up this issue.

**Comment by khyatisoneji** [comment]
@ragesoss For timeline, we calculate the `weeksBeforeTimeline` variable based on course's `start` and `timeline_start`.
How would we calculate `weeksBeforeTimeline` for Milestones component?

**Comment by ragesoss** [comment]
@khyatisoneji weeksBeforeTimeline would be calculated the same way, and that's probably the first thing that could be moved into a selector, so that the result gets cached and shared between Milestons and Timeline.

In Milestone, we don't need to push any extra blocks, we just need to increment the week count.

Currently, Milestone is using week.order for the numbering, while Timeline starts the numbering with `1` plus the weeks before the timeline, and then increments an extra time for each empty week.

Here's an example course where this is a problem: https://dashboard.wikiedu.org/courses/Princeton_University/Linguistic_Universals_and_Language_Diversity_(Spring_2019)

The Milestones section lists "This week, everyone should create a Wikipedia account." in Week 1, but the Timeline shows that as Week 2 (which is the first week of the timeline, because there's one empty week before the start of the timeline). The last milestone, "Once you receive your grade and final feedback...", is week 14 in Milestones but week 16 in the Timeline, because there's also an empty week (Week 7) in the middle.

**Comment by khyatisoneji** [comment]
@ragesoss I am not sure if I completely understand the changes that are required to achieve this.
I get the issue with weeksBeforeTimeline not being part of Milestones and the relevant issue in the course link that you provided.

If I got this correctly,
We need to calculate weeksBeforeTimeline by passing in course start and then do:
```
const weeksBeforeTimeline =CourseDateUtils.weeksBeforeTimeline(course); // course will have start and timeline_start function
```

And then in component, we need to do:
```
<p>Week {index + weeksBeforeTimeline + 1} {completionNote}</p>
```

instead of current:
```
 <p>Week {week.order} {completionNote}</p>
```

Please inform about what needs to be done. Thanks :)

**Comment by ragesoss** [comment]
Yes, I think you have that correctly. The one missing part is the the index, not only increments as you loop through the weeks, but may also increment for an empty week, which is not represented in the `weeks` array, and is handled in Timeline by the `while` statement for weekMeetings that equal `()`, which means a calendar week with no meetings.

**Comment by ragesoss** [comment]
I think this is fixed now, via duplicate #4621
